src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
 
 #include <config.h>
 #include <fastpath/fastpath.h>
-#ifdef CONFIG_KERNEL_MCS
-#include <object/reply.h>
-#endif
 
 #ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
 #include <benchmark/benchmark_track.h>
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_SIGNAL_FASTPATH
+#ifdef CONFIG_ARCH_ARM
+static inline
+FORCE_INLINE
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    word_t fault_type;
+    sched_context_t *sc = NULL;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
+
+    /* Check it's a notification */
+    if (unlikely(!cap_capType_equals(cap, cap_notification_cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to send to this cap */
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+    tcb_t *dest;
+
+    if (ntfnState == NtfnState_Waiting) {
+        /* get the destination thread */
+        dest = TCB_PTR(notification_ptr_get_ntfnQueue_head(ntfnPtr));
+    } else {
+        /* get the bound tcb */
+        dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+    }
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Signal to higher prio thread is NOT fastpathed*/
+            if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
```

Comment:
This is also missing from the start of the NtfnState_Waiting case.

Reply:
I also wonder if anything cross-node should be slow path. There's also the cases where the notification-bound SC is for a different core or requires migrating the dest TCB across cores. This should probably all just go slowpath for now.

Reply:
That's all avoided by these priority checks (when corrected): It assures that the target of the notification doesn't need to be scheduled.

Reply:
`migrateTCB()` is called as part of `schedContext_donate()` which happens when the thread becomes unblocked by the signal. Whether a TCB needs to be migrated shouldn't be affected by priorities, only if it's affinity is not equal to the SC affinity of the donated SC in the case where the thread is passive and the Notification object has an SC bound.

Reply:
Only thing `migrateTCB` does is lazy FPU switching. That seems like a bad idea with SMP enabled to be honest, it's altogether a performance sink and way too sneaky.

With this in mind I agree it's a good idea to make cross-node slow path.

Reply:
I'm struggling to understand this path a little. Would be great if I could clarify my understanding with you guys.  

The call to `migrateTCB()` checks if the `to` thread that  the scheduling context is being donated to currently has its FPU state in the FPU of the CPU it is associated with.  

If it does, then there is check whether this thread's associated CPU is the current one, in which case we call `switchLocalFpuOwner()`, or a different one, in which case we call `doRemoteswitchFpuOwner`

Both of these functions seem to save the current FPU state back to the thread's TCB, disable the FPU and then give up ownership of it, which I think is associated with the lazy FPU switching @Indanz mentioned. The difference between the two  is that the latter does a remote call to a different CPU whereas the former does it for the current. 

Basically, as I understand it, what this code is doing is that it ensures the thread's TCB contains its up to date FPU state in case it needs to migrate to another CPU as a result of the scheduling context having a different CPU associated to it. Is it true to say that if the tcbAffinity of dest is the same as the core of the SC, we don't need to worry about saving FPU state and giving up ownership, as there would be no need for the thread to migrate between CPUs?

So the change to the implementation would be something like 
```
if (dest->tcbAffinity != sc->scCore) {
    slowpath()
}
```

and instead of using `schedContext_donate()` as I currently do, I would use a custom one that avoids the migrateTCB call (as well as getting rid of the `if (from)` conditional block since this case is also slowpathed )?

Reply:
I think you understood everything correctly.

> Is it true to say that if the tcbAffinity of dest is the same as the scCore of the SC, we don't need to worry about saving FPU state and giving up ownership, as there would be no need for the thread to migrate between CPUs?

Only if scCore of the SC is equal to `getCurrentCPUIndex`, as that is what `switchFpuOwner` checks against, but that should be true when donating a scheduling context (all this is triggered via `maybeDonateSchedContext`). 

However, keep in mind that SMP code is not verified and any subtle corner case you missed will probably go undetected.

Currently `migrateTCB` is called unconditionally and always migrates a TCB, even if the old core is equal to the new one. Is this intentional or a bug? It seems like a bug, as it means that if a passive task uses fpu and is called by tasks not using fpus, fpu switching will be done all the time while there are no other fpu users.

Reply:
> Only if scCore of the SC is equal to getCurrentCPUIndex, as that is what switchFpuOwner checks against, but that should be true when donating a scheduling context (all this is triggered via maybeDonateSchedContext).

I'm not sure about this because the donation isn't from the current thread -> destination thread, it's the notification's bound SC being donated to the destination thread. Is there anything that asserts that this will always run on the same CPU as the signalling thread? 

I'm also a bit uncertain about the "if scCore of the SC is equal to `getCurrentCPUIndex`" part. Provided that `dest` is already set up to run on the CPU of the scheduling context, is there any difference whether this is the current CPU or not? `switchFPUOwner` seems to do the FPU save regardless, it's just that for the current CPU it does a normal function call whereas for a different one it does a remote call. I'm not sure why it would make a difference if scCore is different to the current CPU unless we were trying to also fastpath the switchLocalFpuOwner case but not the `doRemoteswitchFpuOwner`.

Reply:
> the donation isn't from the current thread -> destination thread, it's the notification's bound SC being donated to the destination thread.

Of course, signals, not IPC. My excuse is that I had COVID when writing that, still recovering.

> Is there anything that asserts that this will always run on the same CPU as the signalling thread?

None whatsoever, any task may signal any other task on any CPU core. And like Kent said:

> if it's affinity is not equal to the SC affinity of the donated SC in the case where the thread is passive and the Notification object has an SC bound.

The corner case is a passive thread which last ran on a CPU core other than the Notification's bound SC, because of a recent IPC call. This will basically happen all the time on SMP for passive servers which also have notifications.

> Provided that dest is already set up to run on the CPU of the scheduling context, is there any difference whether this is the current CPU or not?

No, there isn't. But like I said, current code doesn't check if the task actually needs to be migrated or not, it just does it. And that doesn't seem something you want to do in your fast path, even when easy to add and possible to do.

I think the current slow path should be streamlined a bit first before implementing a subset in a fast path: Merge `Arch_migrateTCB` with `migrateTCB` and only migrate if newcore != oldcore. Current behaviour can give very bad performance for no good reason it seems, because of the unnecessary FPU switching.

Lazy FPU loading seems to work badly in combination with SMP and passive threads, as TCB migration should be fast.

Reply:
Am I right in concluding that for this PR, we either need to resolve the first half of https://github.com/seL4/seL4/issues/901 or take the slowpath if we're on SMP and the thread to be awoken is passive and the notification has a bound SC that isn't already bound to a thread?

Reply:
Yes, but I think the slowpath condition is narrower than that. The FPU state save (and hence the slowpath condition) only needs to happen if the signalled thread has its FPU state currently in the FPU, as can be seen in the following snippet from Arch_MigrateTCB

```
  if (nativeThreadUsingFPU(thread)) {
      switchFpuOwner(NULL, thread->tcbAffinity);
  }
```
If this is not the case, then the slowpath doesn't seem to touch the FPU. What I currently have in the fastpath does what I think the right behavior is, which is going to the slowpath if the signaled (passive) thread is using the FPU of its previous core but needs to be migrated to a different one, but this is not equivalent with the slowpath behavior and is more like what I think would happen if #901 is resolved.

Reply:
It might be better to make this PR consistent with the slowpath as it is now so that it can be merged, and the SMP issues be addressed as a follow on, otherwise this could keep dragging out...

Reply:
Okay, will do.